### PR TITLE
#163 Kafka connect crud

### PR DESCRIPTION
**Pull request by RamazanYapparov**
PR for issue #163 

Following endpoints have been implemented:

Get all connector names:
`GET /api/clusters/{clusterName}`

Create new connector:
`POST /api/clusters/{clusterName}`

Get connector information:
`GET /api/clusters/{clusterName}/connectors/{connectorName}`

Delete connector:
`DELETE /api/clusters/{clusterName}/connectors/{connectorName}`

Get connector configuration:
`GET /api/clusters/{clusterName}/connectors/{connectorName}/config`

Get connector status:
`GET /api/clusters/{clusterName}/connectors/{connectorName}/status`

Restart connector:
`POST /api/clusters/{clusterName}/connectors/{connectorName}/restart`

Pause connector:
`PUT /api/clusters/{clusterName}/connectors/{connectorName}/pause`

Resume connector:
`PUT /api/clusters/{clusterName}/connectors/{connectorName}/resume`

Update connector configuration
`PUT /api/clusters/{clusterName}/connectors/{connectorName}/config`

Get connectors tasks:
`GET /api/clusters/{clusterName}/connectors/{connectorName}/tasks`

Get task status:
`GET /api/clusters/{clusterName}/connectors/{connectorName}/tasks/{taskId}/status`

Restart task:
`POST /api/clusters/{clusterName}/connectors/{connectorName}/tasks/{taskId}/restart`

Get connector plugins:
`GET /api/clusters/{clusterName}/connector-plugins`

Validate connector plugin configuration:
`PUT /api/clusters/{clusterName}/connector/{pluginName}/config/validate`

### CODE REVIEW

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    public Mono<Void> resumeConnector(String clusterName, String connectorName) {
+        return clustersStorage.getClusterByName(clusterName)
+                .map(cluster -> webClient.put()
+                        .uri(cluster.getKafkaConnect() + "/connectors/{connectorName}/resume", connectorName)
+                        .retrieve()
+                        .bodyToMono(Void.class)
+                        .retryWhen(
+                                Retry.onlyIf(e -> e.exception() instanceof WebClientResponseException.Conflict)
+                                        .retryMax(MAX_RETRIES))
```
Comment (IldarAlmakaev) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java:158]: Let's move this duplicate retry logic to a method/variable

File: kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java [hunk 2]
```diff
@@ -1,35 +1,56 @@
...
-@Testcontainers
 public abstract class AbstractBaseTest {
-    @Container
-    public static KafkaContainer kafka = new KafkaContainer(DockerImageName.parse("confluentinc/cp-kafka:5.2.1"))
-            .withNetwork(Network.SHARED);
-    @Container
-    public static SchemaRegistryContainer schemaRegistry = new SchemaRegistryContainer("5.2.1")
-            .withKafka(kafka)
-            .dependsOn(kafka);
+
```
Comment (IldarAlmakaev) [inline kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java:21]: Why did you remove managing test containers lifecycle using `@Container`?

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    public Mono<Connector> getConnector(String clusterName, String connectorName) {
+        return clustersStorage.getClusterByName(clusterName)
+                .map(cluster -> webClient.get()
+                        .uri(cluster.getKafkaConnect() + "/connectors/{connectorName}", connectorName)
+                        .retrieve()
+                        .bodyToMono(Connector.class)
+                        .retryWhen(
+                                Retry.onlyIf(e -> e.exception() instanceof WebClientResponseException.Conflict)
+                                        .retryMax(MAX_RETRIES))
+                        .doOnError(log::error))
```
Comment (IldarAlmakaev) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java:66]: What happens if retry logic exceeds the **MAX_RETRIES**?
Do we handle this case and what response do we return to clients?

**Review by IldarAlmakaev — APPROVED** [review]
Left some questions for clarification.
Overall, looks good to me

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml [hunk 4]
```diff
@@ -641,6 +641,380 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Connector'
+        409:
+          description: rebalance is in progress
+
+
+  /api/clusters/{clusterName}/connectors/{connectorName}/status:
+    get:
```
Comment (germanosin) [inline kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:794]: Why we need this as a separate endpoint? Could this be included in connector body?

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml [hunk 5]
```diff
@@ -641,6 +641,380 @@ paths:
...
+      responses:
+        200:
+          description: OK
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ConnectorStatus'
+
+  /api/clusters/{clusterName}/connectors/{connectorName}/restart:
+    post:
```
Comment (germanosin) [inline kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:819]: 1. Why is it post, but pause/resume are put?
2. Imho actions in rest looks bad.

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml [hunk 6]
```diff
@@ -641,6 +641,380 @@ paths:
...
+          description: OK
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/ConnectorTask'
+
+  /api/clusters/{clusterName}/connectors/{connectorName}/tasks/{taskId}/status:
+    get:
```
Comment (germanosin) [inline kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:911]: Could we include task status to task object?

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml [hunk 7]
```diff
@@ -641,6 +641,380 @@ paths:
...
+      responses:
+        200:
+          description: OK
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TaskStatus'
+
+  /api/clusters/{clusterName}/connectors/{connectorName}/tasks/{taskId}/restart:
+    post:
```
Comment (germanosin) [inline kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:941]: same here about post/put/actions

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml [hunk 8]
```diff
@@ -641,6 +641,380 @@ paths:
         404:
           description: Not Found
 
+  /api/clusters/{clusterName}/connectors:
```
Comment (germanosin) [inline kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:644]: One cluster might be used by multiple instances of kafka connect, and we should be able to manage multiple from rest per cluster

**Review by germanosin — CHANGES_REQUESTED** [review]

File: kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java (hunk 2 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java:21]: Because using `@Container` the containers did restart for every test class, but application properties were not updated with new container addresses

Comment (IldarAlmakaev) [reply kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java:21]: Hmm...Ok, I'll check

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java (hunk 1 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java]: Fixed, thanks

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java (hunk 3 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java]: Added handling, thanks

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml (hunk 4 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:814]: Merged into one, thanks

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml (hunk 5 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml]: Updated, now it's all `POST` and url is `connectors/{name}/action/{action}`

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml (hunk 6 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:879]: Merged into one, thanks

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml (hunk 7 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:911]: Now it's all post

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml (hunk 8 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml]: Updated, thanks

File: docker/kafka-clusters-only.yaml [hunk 9]
```diff
@@ -88,6 +88,33 @@ services:
     ports:
     - 8081:8081
 
+  kafka-connect0:
+    image: confluentinc/cp-kafka-connect:5.1.0
+    ports:
+      - 8083:8083
+    depends_on:
+      - zookeeper0
+      - kafka0
```
Comment (IldarAlmakaev) [inline docker/kafka-clusters-only.yaml:97]: FYI, Kafka Connect doesn't depend on Zookeeper

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml [hunk 10]
```diff
@@ -641,6 +642,364 @@ paths:
         404:
           description: Not Found
 
+  /api/clusters/{clusterName}/connect:
```
Comment (germanosin) [inline kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml:645]: Lets use plural style in rest paths

File: kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+tags:
+  - name: /connect
+servers:
+  - url: /localhost
+
+paths:
+  /connectors:
+    get:
+      tags:
+        - /connect
```
Comment (germanosin) [inline kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml:20]: tags looks strange, why is it like a path?

File: kafka-ui-contract/pom.xml [hunk 12]
```diff
@@ -33,6 +33,11 @@
                     <artifactId>jackson-databind-nullable</artifactId>
                     <version>${jackson-databind-nullable.version}</version>
                 </dependency>
+                <dependency>
+                    <groupId>com.google.code.findbugs</groupId>
+                    <artifactId>jsr305</artifactId>
+                    <version>3.0.2</version>
```
Comment (germanosin) [inline kafka-ui-contract/pom.xml:39]: why do we need this in runtime?

File: kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java (hunk 2 above)
Comment (germanosin) [reply kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java:21]: @IldarAlmakaev 

> Containers declared as static fields will be shared between test methods. They will be started only once before any test method is executed and stopped after the last test method has executed. Containers declared as instance fields will be started and stopped for every test method.

I would prefer to do it not in static init but in Initializer initialize method and add shutdown hook

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+                .flatMapMany(connect -> withRetryOnConflict(
+                        KafkaConnectClients.withBaseUrl(connect).getConnectors()
+                        ).doOnError(log::error)
+                );
+    }
+
+    public Mono<Connector> createConnector(String clusterName, String connectName, Mono<NewConnector> connector) {
+        return getConnectAddress(clusterName, connectName)
+                .flatMap(connect ->
+                        connector.cache()
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java:59]: Why do you cache request? may be it's better to flatmap it?

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+                                .map(kafkaConnectMapper::toClient)
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java:60]: Is it possible to cache clients and do not recreate them on each request?

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+                                .findFirst()
+                                .map(KafkaConnectCluster::getAddress)
+                )
+                .flatMap(connect -> connect
+                        .map(Mono::just)
+                        .orElse(Mono.error(new NotFoundException("No such connect cluster")))
+                );
+    }
+
+    private <T> Mono<T> withRetryOnConflict(Mono<T> publisher) {
```
Comment (germanosin) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java:201]: you call this method on every call with KafkaConnectClients, maybe it make sense to do it there?

**Review by germanosin — CHANGES_REQUESTED** [review]

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/client/KafkaConnectClients.java [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class KafkaConnectClients {
+
+    private static final Map<String, ConnectApi> CACHE = new ConcurrentHashMap<>();
+
+    public static ConnectApi withBaseUrl(String basePath) {
+        return CACHE.computeIfAbsent(basePath, path -> new ConnectApi(new ApiClient().setBasePath(path)));
```
Comment (alexeyzavyalov) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/client/KafkaConnectClients.java:14]: I think it's better to expire items in the cache after some time. You can use existing cache implementation in this case

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/rest/KafkaConnectRestController.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Override
+    public Mono<ResponseEntity<Flux<String>>> getConnectors(String clusterName, String connectName, ServerWebExchange exchange) {
+        Flux<String> connectors = kafkaConnectService.getConnectors(clusterName, connectName);
+        return Mono.just(ResponseEntity.ok(connectors));
+    }
+
+    @Override
+    public Mono<ResponseEntity<Connector>> createConnector(String clusterName, String connectName, @Valid Mono<NewConnector> connector, ServerWebExchange exchange) {
```
Comment (alexeyzavyalov) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/rest/KafkaConnectRestController.java:35]: Do we need to wrap the response in ResponseEntity, because it's always OK in your code? as I remember we can tell open API generator do not wrap the response into ResponseEntity

File: kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml [hunk 18]
```diff
@@ -0,0 +1,495 @@
...
+            schema:
+              $ref: '#/components/schemas/NewConnector'
+      responses:
+        200:
+          description: OK
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Connector'
+        400:
```
Comment (alexeyzavyalov) [inline kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml:49]: Do we return 400 in some cases? I think it's only 200 or 500

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+
+    public Mono<Connector> getConnector(String clusterName, String connectName, String connectorName) {
+        return getConnectAddress(clusterName, connectName)
+                .flatMap(connect -> withRetryOnConflict(KafkaConnectClients.withBaseUrl(connect).getConnector(connectorName))
+                        .map(kafkaConnectMapper::fromClient)
+                        .flatMap(connector ->
+                                withRetryOnConflict(KafkaConnectClients.withBaseUrl(connect).getConnectorStatus(connector.getName()))
+                                        .map(connectorStatus -> {
+                                            var status = connectorStatus.getConnector();
+                                            return connector.status(kafkaConnectMapper.fromClient(status));
```
Comment (alexeyzavyalov) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java:80]: Am I right that here you mutate the Connector object?

File: kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java (hunk 2 above)
Comment (IldarAlmakaev) [reply kafka-ui-api/src/test/java/com/provectus/kafka/ui/AbstractBaseTest.java:21]: and what is the difference between TestContainer's approach and Initializer's one, @germanosin?
As I understand, we would like to create a test container as a singleton for all our tests not only for class level tests.

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/client/KafkaConnectClients.java (hunk 16 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/client/KafkaConnectClients.java:14]: Do we really need expiration here?
The client instances are limited to number of connect clusters we provide during startup time

File: kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml (hunk 18 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml:49]: This is api for [Kafka Connect API](https://docs.confluent.io/5.2.4/connect/references/restapi.html) that returns both 400 and 500.
I'll add 500 here too, thanks

File: kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml (hunk 10 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/src/main/resources/swagger/kafka-ui-api.yaml]: Fixed

File: kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml (hunk 11 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml:20]: What do you suggest for this one?

File: kafka-ui-contract/pom.xml (hunk 12 above)
Comment (RamazanYapparov) [reply kafka-ui-contract/pom.xml:39]: Changed scope to `provided` thanks

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java (hunk 13 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java]: It was needed in earlier version of this service, removed

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java (hunk 15 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java]: Fixed

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java (hunk 19 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/cluster/service/KafkaConnectService.java]: Thanks, fixed

File: kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml (hunk 11 above)
Comment (IldarAlmakaev) [reply kafka-ui-contract/src/main/resources/swagger/kafka-connect-api.yaml:20]: It shouldn't have a slash. That's it :)

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/rest/KafkaConnectRestController.java (hunk 17 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/rest/KafkaConnectRestController.java:35]: We use `ResponseEntity` classes in other parts of the code

File: kafka-ui-api/src/main/resources/application-local.yml [hunk 20]
```diff
@@ -5,13 +5,17 @@ kafka:
       bootstrapServers: localhost:9093
       zookeeper: localhost:2181
       schemaRegistry: http://localhost:8081
-#      schemaNameTemplate: "%s-value"
+      kafkaConnect:
+        - name: first
+          address: http://localhost:8083
+      #      schemaNameTemplate: "%s-value"
       jmxPort: 9997
```
Comment (IldarAlmakaev) [inline kafka-ui-api/src/main/resources/application-local.yml:12]: Please review `schemaNameTemplate`

**Review by IldarAlmakaev — APPROVED** [review]

File: kafka-ui-api/src/main/resources/application-local.yml (hunk 20 above)
Comment (RamazanYapparov) [reply kafka-ui-api/src/main/resources/application-local.yml:11]: Fixed, thanks

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/rest/KafkaConnectRestController.java (hunk 17 above)
Comment (alexeyzavyalov) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/rest/KafkaConnectRestController.java:35]: I see, but the purpose of using this is not clear for me...

**Review by germanosin — APPROVED** [review]
